pyfar/dsp/filter/frequency_weighting.py

```diff
@@ -0,0 +1,370 @@
+
+    poles_angular = np.array(poles) * -2 * np.pi
+    return sps.bilinear_zpk(zeros, poles_angular, 1, fs)
+
+
+def _x2zpk(x: np.ndarray,
+           target_weighting: Literal["A", "C"],
+           ) -> tuple[np.ndarray, np.ndarray, float]:
+    """
+    Reobtain z, p and k from x (the 1-D array of optimizer inputs).
+    Adds back the fixed zeros at 0 Hz that are excluded from optimization.
+    """
+    fixed_zeros = [1, 1, 1, 1] if target_weighting == "A" else [1, 1]
+    zeros = np.concat([fixed_zeros, x[0:2]])
+    return (zeros, x[2:-1], x[-1])
+
+
+
+def _get_error_margins(f: np.ndarray):
+    """
+    Returns the upper and lower error margin in dB for every frequency in the
+    input array for the A and C weighting according to IEC/DIN-EN 61672-1.
+    Margins are taken from table 3 in the standard.
+    """
+    upper = np.zeros_like(f)
+    upper[f <= 10] = 3.0
+    upper[(f > 10) & (f <= 12.5)] = 2.5
+    upper[(f > 12.5) & (f <= 25)] = 2.0
+    upper[(f > 25) & (f <= 31.5)] = 1.5
+    upper[(f > 31.5) & (f <= 800)] = 1.0
+    upper[(f > 800) & (f < 1250)] = 0.7
+    upper[(f >= 1250) & (f < 5000)] = 1.0
+    upper[(f >= 5000) & (f < 10e3)] = 1.5
+    upper[(f >= 10e3) & (f < 16e3)] = 2.0
+    upper[(f >= 16e3) & (f < 20e3)] = 2.5
+    upper[f >= 20e3] = 3.0
+
+    lower = np.zeros_like(f)
+    lower[f <= 12.5] = -np.inf
+    lower[(f > 12.5) & (f <= 16)] = -4.0
+    lower[(f > 16) & (f <= 20)] = -2.0
+    lower[(f > 20) & (f <= 31.5)] = -1.5
+    lower[(f > 31.5) & (f <= 800)] = -1.0
+    lower[(f > 800) & (f < 1250)] = -0.7
+    lower[(f >= 1250) & (f < 5000)] = -1.0
+    lower[(f >= 5000) & (f < 6300)] = -1.5
+    lower[(f >= 6300) & (f < 8000)] = -2.0
+    lower[(f >= 8000) & (f < 10e3)] = -2.5
+    lower[(f >= 10e3) & (f < 12500)] = -3.0
+    lower[(f >= 12500) & (f < 16e3)] = -5.0
+    lower[(f >= 16e3) & (f < 20e3)] = -16.0
+    lower[f >= 20e3] = -np.inf
+
+    return upper, lower
+
+
+def _check_filter(sampling_rate: float,
+                  sos: np.ndarray,
+                  weighting: Literal["A", "C"],
+                  ) -> tuple[bool, float, float]:
+    """
+    Checks whether the filter is class 1 compliant according to the norm and
+    provides simple error statistics.
+    """
+    test_freqs = np.logspace(1, np.log10(sampling_rate / 2), 100_000)
+
+    # calculate the magnitude differences between the filter response
+    # and the target weighting
+    z, p, k = sps.sos2zpk(sos)
+    _, freq_resp = sps.freqz_zpk(z, p, k, test_freqs, fs=sampling_rate)
+    mags = np.abs(freq_resp)
+    mags_dB = 10 * np.log10(mags**2)
+    mags_target = pfc.frequency_weighting_curve(weighting, test_freqs)
+    mag_diffs = mags_dB - mags_target
+
+    upper, lower = _get_error_margins(test_freqs)
+    is_class_1 = np.all(mag_diffs < upper) and np.all(mag_diffs >= lower)
+    max_diff = np.max(np.abs(mag_diffs))
+    mean_diff = np.mean(np.abs(mag_diffs))
+    return is_class_1, max_diff, mean_diff
```

Comment:
Maybe this function could also modified to allow user access to the error margins similar to the tolerance functions for fractional octave filters?